### ISSUE TITLE
AllenNLP Machine Comprehension Model is generating different outputs at different runs

### ISSUE THREAD

**Issue by ghoshmithun** [allenai/allennlp#1749]
I have created a server using the archived model. It is generating different outputs at different runs. But it should not be the case
 Model : => "https://s3-us-west-2.amazonaws.com/allennlp/models/bidaf-model-2017.09.15-charpad.tar.gz"
Objective is to build a question answering machine given a paragraph( comprehend the paragraph)

**To Reproduce (Python code) (Below are the package requirements)**
      
       from allennlp.models import load_archive
       from allennlp.data import DatasetReader
       from allennlp.data.dataset import Batch

    archive = load_archive("bidaf-model-2017.09.15-charpad.tar.gz")
    #archive.model
    #type(archive.model)

    config = archive.config.duplicate()
    dataset_reader_params = config["dataset_reader"]
    dataset_reader = DatasetReader.from_params(dataset_reader_params)

    model = archive.model
    question="How is GloVe different from word2vec?"
    passage="Both models learn geometrical encodings (vectors) of words from their co-occurrence information (how frequently they appear together in large text corpora). They differ in that word2vec is a predictive model, whereas GloVe is a count-based model. See this paper for more on the distinctions between these two approaches: http://clic.cimec.unitn.it/marco.... Predictive models learn their vectors in order to improve their predictive ability of Loss(target word | context words; Vectors), i.e. the loss of predicting the target words from the context words given the vector representations. In word2vec, this is cast as a feed-forward neural network and optimized as such using SGD, etc. Count-based models learn their vectors by essentially doing dimensionality reduction on the co-occurrence counts matrix. They first construct a large matrix of (words x context) co-occurrence information, i.e. for each word (the rows), you count how frequently we see this word in some context (the columns) in a large corpus.  The number of contexts is of course large, since it is essentially combinatorial in size. So then they factorize this matrix to yield a lower-dimensional (word x features) matrix, where each row now yields a vector representation for each word. In general, this is done by minimizing a reconstruction loss which tries to find the lower-dimensional representations which can explain most of the variance in the high-dimensional data. In the specific case of GloVe, the counts matrix is preprocessed by normalizing the counts and log-smoothing them. This turns out to be A Good Thing in terms of the quality of the learned representations. However, as pointed out, when we control for all the training hyper-parameters, the embeddings generated using the two methods tend to perform very similarly in downstream NLP tasks. The additional benefits of GloVe over word2vec is that it is easier to parallelize the implementation which means it is easier to train over more data, which, with these models, is always A Good Thing."
     for i in range(5):
          instance = dataset_reader.text_to_instance( question,passage ,char_spans=[(6, 10)])
          instances = [instance]
          dataset = Batch(instances)
          dataset.index_instances(model.vocab)

          cuda_device = model._get_prediction_device()
          model_input = dataset.as_tensor_dict(cuda_device=cuda_device)

          outputs = model(**model_input)
          response = outputs['best_span_str']
          print(response)

In above case model is taking a question and a passage but it is generating different output at different run . Ideally it should return unique output at each case. I am really puzzled 

System:
 - OS: OS 11.2
_**cat requirements.txt**_ 
alabaster==0.7.11
allennlp==0.6.0
atomicwrites==1.1.5
attrs==18.1.0
awscli==1.15.81
Babel==2.6.0
boto3==1.7.62
botocore==1.10.80
certifi==2018.8.13
cffi==1.11.2
chardet==3.0.4
click==6.7
colorama==0.3.9
conllu==0.11
cookies==2.2.1
cymem==1.31.2
cytoolz==0.9.0.1
dill==0.2.8.2
docutils==0.14
editdistance==0.4
en-core-web-sm==2.0.0
flaky==3.4.0
Flask==0.12.1
Flask-Cors==3.0.3
gevent==1.3.5
greenlet==0.4.14
h5py==2.8.0
idna==2.7
imagesize==1.0.0
itsdangerous==0.24
Jinja2==2.10
jmespath==0.9.3
jsonnet==0.10.0
MarkupSafe==1.0
more-itertools==4.3.0
msgpack==0.5.6
msgpack-numpy==0.4.3.1
murmurhash==0.28.0
nltk==3.3
numpy==1.15.0
numpydoc==0.8.0
overrides==1.9
packaging==17.1
parsimonious==0.8.0
plac==0.9.6
pluggy==0.7.1
preshed==1.0.1
protobuf==3.6.1
py==1.5.4
pyasn1==0.4.4
pycparser==2.18
Pygments==2.2.0
pyparsing==2.2.0
pytest==3.7.2
python-dateutil==2.7.3
pytz==2017.3
PyYAML==3.13
regex==2017.4.5
requests==2.19.1
responses==0.9.0
rsa==3.4.2
s3transfer==0.1.13
scikit-learn==0.19.2
scipy==1.1.0
six==1.11.0
snowballstemmer==1.2.1
spacy==2.0.12
Sphinx==1.7.7
sphinxcontrib-websupport==1.1.0
tensorboardX==1.2
thinc==6.10.3
toolz==0.9.0
torch==0.4.1
tqdm==4.24.0
typing==3.6.4
ujson==1.35
Unidecode==1.0.22
urllib3==1.23
Werkzeug==0.14.1
wrapt==1.10.11

**Comment by matt-gardner** [comment]
I can replicate this.  It turns out this is expected, though.  There's a small amount of non-determinism in the floating point operations done by tensor libraries.  When you couple that with a question that doesn't have a clear answer in the paragraph, you can get the "best" answer bouncing around between the various modes of the answer distribution.

**Comment by ghoshmithun** [comment]
Thank you Matt. But while I run the same problem in "http://demo.allennlp.org/machine-comprehension", it seems to provide consistent solution or same string. Am I missing anything or what are the modification to be made in the code. Can you please guide me to the S3 location if you have model which has been trained later. The Model archive is quite old. Greatly appreciated your prompt response

**Comment by matt-gardner** [comment]
I tried putting the archive loading into the loop, so the model is loaded from scratch each time, and I still get inconsistent results on my (mac) laptop.  I also tried running the demo a whole bunch of times, and I got very consistent results each time.  I'm guessing the difference is in the underlying tensor libraries that are being used - maybe the one that's compiled on my mac just has more non-determinism than the one that's compiled on the linux machine running the demo.  There's unfortunately not much that we can do about this.
